### PR TITLE
Fire a transfer queued event when a transfer is queued in Graphsync

### PR DESCRIPTION
**Pull request by aarshkshah1992**
Depends ON https://github.com/ipfs/go-graphsync/pull/172.

### TODO
- [ ] Merge and release Graphsync PR.
- [ ] Tests.

### CODE REVIEW

File: transport/graphsync/graphsync.go [hunk 1]
```diff
@@ -448,6 +449,51 @@ func (t *Transport) gsOutgoingBlockHook(p peer.ID, request graphsync.RequestData
 	}
 }
 
+// gsReqRecdHook is called when graphsync receives an incoming request for data
+func (t *Transport) gsReqQueuedHook(p peer.ID, request graphsync.RequestData) {
+	msg, err := extension.GetTransferData(request, t.supportedExtensions)
+	if err != nil {
+		return
```
Comment (dirkmc) [inline transport/graphsync/graphsync.go:456]: Let's log the error here

File: transport/graphsync/graphsync.go [hunk 2]
```diff
@@ -448,6 +449,51 @@ func (t *Transport) gsOutgoingBlockHook(p peer.ID, request graphsync.RequestData
 	}
 }
 
+// gsReqRecdHook is called when graphsync receives an incoming request for data
```
Comment (dirkmc) [inline transport/graphsync/graphsync.go:452]:
```suggestion
// gsReqQueuedHook is called when graphsync enqueues an incoming request for data
```

File: transport/graphsync/graphsync.go [hunk 3]
```diff
@@ -448,6 +449,51 @@ func (t *Transport) gsOutgoingBlockHook(p peer.ID, request graphsync.RequestData
...
+		// initiated a pull
+		chid = datatransfer.ChannelID{ID: msg.TransferID(), Initiator: p, Responder: t.peerID}
+		// Lock the channel for the duration of this method
+		ch = t.trackDTChannel(chid)
+		ch.lk.Lock()
+		defer ch.lk.Unlock()
+
+		request := msg.(datatransfer.Request)
+		if request.IsRestart() || request.IsNew() {
+			log.Infof("channelID=%s, pull request queued, req=%+v", chid, request)
```
Comment (dirkmc) [inline transport/graphsync/graphsync.go:476]: For consistency with the other logs in this file
```suggestion
			log.Infof("%s: pull request queued, req=%+v", chid, request)
```

File: transport/graphsync/graphsync.go [hunk 4]
```diff
@@ -448,6 +449,51 @@ func (t *Transport) gsOutgoingBlockHook(p peer.ID, request graphsync.RequestData
...
+		// for data
+		chid = datatransfer.ChannelID{ID: msg.TransferID(), Initiator: t.peerID, Responder: p}
+		// Lock the channel for the duration of this method
+		ch = t.trackDTChannel(chid)
+		ch.lk.Lock()
+		defer ch.lk.Unlock()
+
+		response := msg.(datatransfer.Response)
+		if response.IsNew() || response.IsRestart() {
+			log.Infof("channelID=%s, GS pull request queued in response to our push, req=%+v", chid, request)
```
Comment (dirkmc) [inline transport/graphsync/graphsync.go:491]:
```suggestion
			log.Infof("%s: GS pull request queued in response to our push, req=%+v", chid, request)
```

File: transport/graphsync/graphsync.go [hunk 5]
```diff
@@ -448,6 +449,51 @@ func (t *Transport) gsOutgoingBlockHook(p peer.ID, request graphsync.RequestData
...
+	var chid datatransfer.ChannelID
+	var ch *dtChannel
+	if msg.IsRequest() {
+		// when a data transfer request comes in on graphsync, the remote peer
+		// initiated a pull
+		chid = datatransfer.ChannelID{ID: msg.TransferID(), Initiator: p, Responder: t.peerID}
+		// Lock the channel for the duration of this method
+		ch = t.trackDTChannel(chid)
+		ch.lk.Lock()
+		defer ch.lk.Unlock()
```
Comment (dirkmc) [inline transport/graphsync/graphsync.go:472]: I don't think we need to start tracking the channel at this stage, suggest removing these lines

File: transport/graphsync/graphsync.go [hunk 6]
```diff
@@ -448,6 +449,51 @@ func (t *Transport) gsOutgoingBlockHook(p peer.ID, request graphsync.RequestData
...
+		}
+	} else {
+		// when a data transfer response comes in on graphsync, this node
+		// initiated a push, and the remote peer responded with a request
+		// for data
+		chid = datatransfer.ChannelID{ID: msg.TransferID(), Initiator: t.peerID, Responder: p}
+		// Lock the channel for the duration of this method
+		ch = t.trackDTChannel(chid)
+		ch.lk.Lock()
+		defer ch.lk.Unlock()
```
Comment (dirkmc) [inline transport/graphsync/graphsync.go:487]: I don't think we need to start tracking the channel at this stage, suggest removing these lines

File: channels/channels.go [hunk 7]
```diff
@@ -215,6 +215,10 @@ func (c *Channels) Accept(chid datatransfer.ChannelID) error {
 	return c.send(chid, datatransfer.Accept)
 }
 
+func (c *Channels) TransferRequestQueuedinTransport(chid datatransfer.ChannelID) error {
```
Comment (dirkmc) [inline channels/channels.go:218]: nit: could we change the name to something shorter, eg `TransferRequestQueued`

File: channels/channels.go [hunk 8]
```diff
@@ -215,6 +215,10 @@ func (c *Channels) Accept(chid datatransfer.ChannelID) error {
...
+	return c.send(chid, datatransfer.TransferRequestQueuedinTransport)
```
Comment (dirkmc) [inline channels/channels.go:219]: nit: Same for the event: `TransferRequestQueued`

**Review by dirkmc — APPROVED** [review]
LGTM 👍 
(Just one nit)

**Comment by aarshkshah1992** [comment]
@dirkmc Have addressed your nits.

**Pull request merged**